### ISSUE TITLE
HPO579:  make bucket (mb) BucketAlreadyOwned and then Read timeout

### ISSUE THREAD

**Issue by MonicaLemay** [noobaa/noobaa-core#6907]
This is from HPO defect 579

I get both BucketAlreadyOwned and then Read timeout

First, I get the message BucketAlreadyOwned :

Here is the workflow:

Account s50040 has 10 buckets the ESS and mmdas confirm 10 buckets

```

[root@c83f1-ems hpofs1]# ls -ali s50040
total 262
69671 drwxr-xr-x. 12 50040 50040 4096 Mar 4 16:31 .
3 drwxr-xr-x. 122 root root 262144 Mar 7 18:06 ..
4235040 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b1
4235046 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b10
4235041 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b2
4234998 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b3
4235042 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b4
4235043 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b5
4235044 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b6
4235045 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b7
4234995 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b8
4234902 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b9
```
Now,  let’s make more:

```
 s50040 mb s3://s50040b11
make_bucket failed: s3://s50040b11 An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. You get this error in all AWS regions except US East (N. Virginia) region, us-east-1. In us-east-1 region, you will get 200 OK, but it is no-op (if bucket exists it Amazon S3 will not do anything).

The time it takes is:
real 1m8.956s
user 0m0.504s
sys 0m0.070s
```

Again, another one:
```
s50040 mb s3://s50040b12
make_bucket failed: s3://s50040b12 An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. You get this error in all AWS regions except US East (N. Virginia) region, us-east-1. In us-east-1 region, you will get 200 OK, but it is no-op (if bucket exists it Amazon S3 will not do anything).

real 1m29.830s
user 0m0.505s
sys 0m0.055s
```

I can confirm via mmdas, and on the ESS that the exports get created. It is just very mis-leading to report that it already exists.  I did not capture that here, but I did confirm

By time I get to the 15th bucket, I start to get read timeout.

```
 s50040 mb s3://s50040b15
make_bucket failed: s3://s50040b15 Read timeout on endpoint URL: "http://172.20.100.22/s50040b15"

real 3m4.644s
user 0m0.510s
sys 0m0.064s

```
```
s50040 mb s3://s50040b16
make_bucket failed: s3://s50040b16 Read timeout on endpoint URL: "http://172.20.100.22/s50040b16"

real 3m5.086s
user 0m0.508s
sys 0m0.056s
```

They do get created on Mmdas and on ESS [root@c83f1-infa mb]
```
mmdas export list | grep s50040
s50040b16
s50040b15
s50040b14
s50040b12
s50040b11
s50040b1
s50040b7
s50040b4
s50040b5
s50040b2
s50040b3
s50040b6
s50040b8
s50040b9
s50040b10
```
```
root@c83f1-ems hpofs1]# ls -ail s50040
total 265
69671 drwxr-xr-x. 18 50040 50040 4096 Mar 9 14:25 .
3 drwxr-xr-x. 122 root root 262144 Mar 7 18:06 ..
4235040 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b1
4235046 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b10
174627 drwxr-x---. 2 50040 50040 4096 Mar 9 14:13 s50040b11
174628 drwxr-x---. 2 50040 50040 4096 Mar 9 14:14 s50040b12
174629 drwxr-x---. 2 50040 50040 4096 Mar 9 14:17 s50040b13
174630 drwxr-x---. 2 50040 50040 4096 Mar 9 14:19 s50040b14
174631 drwxr-x---. 2 50040 50040 4096 Mar 9 14:23 s50040b15
174632 drwxr-x---. 2 50040 50040 4096 Mar 9 14:25 s50040b16
4235041 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b2
4234998 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b3
4235042 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b4
4235043 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b5
4235044 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b6
4235045 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b7
4234995 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b8
4234902 drwxr-x---. 2 50040 50040 4096 Mar 4 16:31 s50040b9
[root@c83f1-ems hpofs1]# ls -ail s50040
```
Logs for time out show.
```
 [root@c83f1-infa mb]# oc get pods | grep endpoint | awk '{print $1}
' | xargs -ti oc logs {} | grep s50040b15
oc logs noobaa-endpoint-59f969cd76-2csw5
oc logs noobaa-endpoint-59f969cd76-2xq7k
oc logs noobaa-endpoint-59f969cd76-42ft4
oc logs noobaa-endpoint-59f969cd76-594j7
oc logs noobaa-endpoint-59f969cd76-6j479
oc logs noobaa-endpoint-59f969cd76-6j5g9
oc logs noobaa-endpoint-59f969cd76-6nsxk
oc logs noobaa-endpoint-59f969cd76-7zgp8
oc logs noobaa-endpoint-59f969cd76-96z9s
oc logs noobaa-endpoint-59f969cd76-b5xl9
oc logs noobaa-endpoint-59f969cd76-ccdt9
oc logs noobaa-endpoint-59f969cd76-cfhg5
oc logs noobaa-endpoint-59f969cd76-dtq97
oc logs noobaa-endpoint-59f969cd76-fqwhx
Mar-9 19:21:40.768 [Endpoint/14] [L1] core.endpoint.s3.s3_rest:: S3 REQUEST PUT /s50040b15 op put_bucket request_id l0jy66hs-dexekj-1bmv { host: '172.20.100.22', 'accept-encoding': 'identity', 'user-agent': 'aws-cli/2.3.2 Python/3.8.8 Linux/4.18.0-240.el8.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb', 'x-amz-date': '20220309T192138Z', 'x-amz-content-sha256': '61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1', authorization: 'AWS4-HMAC-SHA256 Credential=eaDxI4TLjlxbwW1jugZC/20220309/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=31f7e36128e06c73e59b229c1b95fd42c9a029b1f20e69dd5b1b3013f2528717', 'content-length': '154' }
Mar-9 19:25:33.734 [Endpoint/14] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR BucketAlreadyOwnedByYouYour previous request to create the named bucket succeeded and you already own it. You get this error in all AWS regions except US East (N. Virginia) region, us-east-1. In us-east-1 region, you will get 200 OK, but it is no-op (if bucket exists it Amazon S3 will not do anything)./s50040b15l0jy66hs-dexekj-1bmv PUT /s50040b15 {"host":"172.20.100.22","accept-encoding":"identity","user-agent":"aws-cli/2.3.2 Python/3.8.8 Linux/4.18.0-240.el8.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20220309T192138Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=eaDxI4TLjlxbwW1jugZC/20220309/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=31f7e36128e06c73e59b229c1b95fd42c9a029b1f20e69dd5b1b3013f2528717","content-length":"154"} [RpcError: BUCKET_ALREADY_OWNED_BY_YOU] { rpc_code: 'BUCKET_ALREADY_OWNED_BY_YOU' }
oc logs noobaa-endpoint-59f969cd76-hdhf2
oc logs noobaa-endpoint-59f969cd76-j9bqv
bucket_info: { name: SENSITIVE-79adbf34f11ab94d, owner_account: { email: SENSITIVE-66d3162d20d7910b, id: '62226d9898554300293755e7' }, namespace: { write_resource: { resource: 'noobaa-s3res-3777249441', path: 's50040/s50040b15' }, read_resources: [ { resource: 'noobaa-s3res-3777249441', path: 's50040/s50040b15' }, [length]: 1 ], should_create_underlying_storage: true }, tag: '', num_objects: { value: 0, last_update: 1646853390000 }, writable: false, storage: { values: { used: 0, used_other: 0, total: 0, free: 0 }, last_update: 1646853390000 }, data: { size: 0, size_reduced: 0, free: 0, available_for_upload: 0, last_update: 1646853390000 }, usage_by_pool: { pools: [ [length]: 0 ], last_update: 1646853390000 }, stats_by_type: [ [length]: 0 ], mode: 'OPTIMAL', bucket_type: 'NAMESPACE', versioning: 'DISABLED', triggers: [ [length]: 0 ] },
write_resource: { resource: { fs_root_path: '/nsfs/noobaa-s3res-3777249441', fs_backend: 'GPFS', id: '622105b0c8206400294c9bcb', name: 'noobaa-s3res-3777249441' }, path: 's50040/s50040b15' },
read_resources: [ { resource: { fs_root_path: '/nsfs/noobaa-s3res-3777249441', fs_backend: 'GPFS', id: '622105b0c8206400294c9bcb', name: 'noobaa-s3res-3777249441' }, path: 's50040/s50040b15' }, [length]: 1 ],
Mar-9 19:23:17.003 [Endpoint/15] [L0] core.sdk.namespace_fs:: NamespaceFS: create_uls fs_account_config: { gid: 50040, uid: 50040, nsfs_only: true, new_buckets_path: 's50040/', backend: 'GPFS', warn_threshold_ms: 100 } new_dir_path: /nsfs/noobaa-s3res-3777249441/s50040/s50040b15
2022-03-09 19:23:17.003298 [PID-15/TID-15] [L1] FS::FSWorker::Begin: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504
2022-03-09 19:23:17.003353 [PID-15/TID-25] [L1] FS::FSWorker::Execute: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504 _uid=50040 _gid=50040 _backend=GPFS
2022-03-09 19:23:17.006042 [PID-15/TID-25] [L1] FS::FSWorker::Execute: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504 took: 2.56592 ms
2022-03-09 19:23:17.006138 [PID-15/TID-15] [L1] FS::FSWorker::OnOK: undefined Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504
Mar-9 19:23:17.006 [Endpoint/15] [L1] core.util.http_utils:: HTTP REPLY EMPTY PUT /s50040b15 {"host":"172.20.100.22","accept-encoding":"identity","user-agent":"aws-cli/2.3.2 Python/3.8.8 Linux/4.18.0-240.el8.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20220309T191936Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=eaDxI4TLjlxbwW1jugZC/20220309/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=9151d31b65016a6b324a4535130cff199ab0f390883c5ee8ecc8c5f9d3aefbbe","content-length":"154"} 200
Mar-9 19:25:46.675 [Endpoint/15] [L0] core.sdk.namespace_fs:: NamespaceFS: create_uls fs_account_config: { gid: 50040, uid: 50040, nsfs_only: true, new_buckets_path: 's50040/', backend: 'GPFS', warn_threshold_ms: 100 } new_dir_path: /nsfs/noobaa-s3res-3777249441/s50040/s50040b15
2022-03-09 19:25:46.676303 [PID-15/TID-15] [L1] FS::FSWorker::Begin: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504
2022-03-09 19:25:46.676456 [PID-15/TID-25] [L1] FS::FSWorker::Execute: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504 _uid=50040 _gid=50040 _backend=GPFS
2022-03-09 19:25:46.678394 [PID-15/TID-25] [L1] FS::FSWorker::Execute: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504 took: 1.7667 ms
2022-03-09 19:25:46.678601 [PID-15/TID-15] [L1] FS::FSWorker::OnError: Mkdir _path=/nsfs/noobaa-s3res-3777249441/s50040/s50040b15 _mode=504 error.Message()=File exists
Mar-9 19:28:04.352 [Endpoint/15] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR BucketAlreadyExistsThe requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again./s50040b15l0jy4utt-lnkno-b4j PUT /s50040b15 {"host":"172.20.100.22","accept-encoding":"identity","user-agent":"aws-cli/2.3.2 Python/3.8.8 Linux/4.18.0-240.el8.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.mb","x-amz-date":"20220309T192036Z","x-amz-content-sha256":"61d056dc66f1882c0f4053be381523a7a28d384abde04fcf5b0021c716bb0ea1","authorization":"AWS4-HMAC-SHA256 Credential=eaDxI4TLjlxbwW1jugZC/20220309/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=bdab9621f58c490f7748ec6aaf3e9c931c9b5dfeed141f05db5c0ab6a1f31be2","content-length":"154"} Error: File exists
oc logs noobaa-endpoint-59f969cd76-jg4jp
oc logs noobaa-endpoint-59f969cd76-k58zj
oc logs noobaa-endpoint-59f969cd76-lg464
oc logs noobaa-endpoint-59f969cd76-lvqdp
oc logs noobaa-endpoint-59f969cd76-ps6l4
oc logs noobaa-endpoint-59f969cd76-qhrtm
oc logs noobaa-endpoint-59f969cd76-rcwnm
oc logs noobaa-endpoint-59f969cd76-shnlt
oc logs noobaa-endpoint-59f969cd76-sxpz8
oc logs noobaa-endpoint-59f969cd76-v6kmk
oc logs noobaa-endpoint-59f969cd76-vsxhz
oc logs noobaa-endpoint-59f969cd76-w4fgr
oc logs noobaa-endpoint-59f969cd76-wj64g
oc logs noobaa-endpoint-59f969cd76-xh6hv
[root@c83f1-infa mb]#
```
Both messages are going to be very confusing to the end user. They may not have the ability to mmdas or look at the storage. They will be able to do s3 ls to see their buckets. Regardless, they should not be getting the above messages.

Making this a sev 1 b/c it makes it very very difficult to scale up my environment for testing.

Logs in box folder.
https://ibm.ent.box.com/folder/145794528783?s=uueh7fp424vxs2bt4ndrnvh7uusgu6tocd
Your environment

POK Bare Metal Real Fast cluster on loan to Real World Cluster
[root@c83f1-infa mb]# noobaa status
INFO[0000] CLI version: 5.9.0
INFO[0000] noobaa-image: noobaa/noobaa-core:5.9-nsfs-fixes-20220303-nsfs
INFO[0000] operator-image: quay.io/rhceph-dev/odf4-mcg-rhel8-operator@sha256:ca66f681fb5435837543995440661e1ef67a5dce85ae3bf91cafe2f0b66b6e5b
INFO[0000] noobaa-db-image: quay.io/rhceph-dev/rhel8-postgresql-12@sha256:b006af9af125adce28546d0ed9d77421abe54182741f6b5e1d31c88dd763580b

[root@c83f1-infa mb]# oc get csv
NAME DISPLAY VERSION REPLACES PHASE
mcg-operator.v4.9.4 NooBaa Operator 4.9.4 mcg-operator.v4.9.3 Succeeded
ocs-operator.v4.9.4 OpenShift Container Storage 4.9.4 ocs-operator.v4.9.3 Succeeded
odf-operator.v4.9.4 OpenShift Data Foundation 4.9.4 odf-operator.v4.9.3 Succeeded
[root@c83f1-infa mb]#

Recently applied odf 4.9.4-7
Steps to reproduce

I am seeing this when I have scaled up to 100 accounts and am trying to increase from 10 buckets per account to 50
Expected behaviour

No error messages in mb

**Comment by romayalon** [comment]
Initial screening:
1. From the endpoint logs (pods: noobaa-endpoint-59f969cd76-fqwhx, noobaa-endpoint-59f969cd76-j9bqv) we can see that the request is sent 3 times. NooBaa returns 200 OK on the bucket creation but it's too late and another request was already sent ( @MonicaLemay  if you will increase the timeout, do you still see these issues? also you can configure the max retries) 
2. In noobaa-endpoint-59f969cd76-fqwhx I see PUT bucket request at 19:21:40.768 and Already exists error at 19:25:33.734.
3. In noobaa-endpoint-59f969cd76-j9bqv I see the mkdir and the successful reply at 19:23:17.006 But I don't see the request of that reply, and at 19:25:46.679 I see already exists error, without a request again.
4. I found there an issue and tried the fix but it still doesn't resolve the defect completely (anyway I'll create a PR for that fix).

**Comment by akmithal** [comment]
Hi @romayalon , Isn't s3 mb a synchronous operation ? If it is, Noobaa should get to know(from app node) that a bucket creation attempt has failed and it should delete the bucket from the database and the filesystem.

If is not a synchronous operation, then such events can always happen where buckets would get created and user wouldn't be able to use them.

What is the fix you are working on ?

**Comment by romayalon** [comment]
@akmithal the situation that I see in logs is not that the first trial failed in NooBaa, it failed on timeout in the client and it wasn't aborted so NooBaa still running all retries including the first one.
I'll keep you updated regarding a solution.

**Comment by romayalon** [comment]
@MonicaLemay I thought you are doing some kind of a loop, so you are just running 40 times `s3 mb` manually for each account of the 100?

**Comment by romayalon** [comment]
@MonicaLemay also if you can try to reproduce this issue and get noobaa core logs that have log messages from the time of the errors it'll be great because in the currently attached logs it starts after a minute of the error.

**Comment by MonicaLemay** [comment]
I probably had logging set to nsfs b/c of defect 6896.  The logs tend to wrap.  Recreates on this are not easy.  This doesn't get hit in FVT. I don't have a specific workflow that hits this.  I get it when trying to set up a scale up of accounts and buckets.  It only happens in system test just by luck.  If I see it again, I will get logs again.

**Comment by romayalon** [comment]
@MonicaLemay Thanks Monica,
I checked again noobaa core logs and found the following error for bucket s50040b17:
```
[32mMar-9 19:29:18.120[35m [WebServer/41] [31m[ERROR][39m core.server.system_services.system_store:: got error on bulk execute error: duplicate key value violates unique constraint "idx_btree_buckets_system_name"
    at Parser.parseErrorMessage (/root/node_modules/[4mnoobaa-core[24m/node_modules/[4mpg-protocol[24m/dist/parser.js:287:98)
    at Parser.handlePacket (/root/node_modules/[4mnoobaa-core[24m/node_modules/[4mpg-protocol[24m/dist/parser.js:126:29)
    at Parser.parse (/root/node_modules/[4mnoobaa-core[24m/node_modules/[4mpg-protocol[24m/dist/parser.js:39:38)
    at Socket.<anonymous> (/root/node_modules/[4mnoobaa-core[24m/node_modules/[4mpg-protocol[24m/dist/index.js:11:42)
[90m    at Socket.emit (events.js:400:28)[39m
[90m    at Socket.emit (domain.js:475:12)[39m
[90m    at addChunk (internal/streams/readable.js:293:12)[39m
[90m    at readableAddChunk (internal/streams/readable.js:267:9)[39m
[90m    at Socket.Readable.push (internal/streams/readable.js:206:10)[39m
[90m    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)[39m {
  length: [33m302[39m,
  severity: [32m'ERROR'[39m,
  code: [32m'23505'[39m,
  detail: [32m"Key ((data ->> 'system'::text), (data ->> 'name'::text))=(62210587c8206400294c9bb7, s50040b17) already exists."[39m,
  hint: [90mundefined[39m,
  position: [90mundefined[39m,
  internalPosition: [90mundefined[39m,
  internalQuery: [90mundefined[39m,
  where: [90mundefined[39m,
  schema: [32m'public'[39m,
  table: [32m'buckets'[39m,
  column: [90mundefined[39m,
  dataType: [90mundefined[39m,
  constraint: [32m'idx_btree_buckets_system_name'[39m,
  file: [32m'nbtinsert.c'[39m,
  line: [32m'570'[39m,
  routine: [32m'_bt_check_unique'[39m
}
```
@MonicaLemay you didn't mention this bucket in the defect but do you remember if you got the same issue for s50040b17 as well? 
This error implies that we passed the bucket existence check and we tried to insert it again to the DB.

**Comment by romayalon** [comment]
@MonicaLemay 
I'm updating that I was able to reproduce it while adding a delay that illustrates the slowness of your system.
I fixed one issue that I saw in the attached logs (the bucket s50040b17 problem I mentioned in the last comment).
But still, it won't fix the slowness, I can't see the reason for it in the provided logs so a reproduction with 'all' debug level is a must.
I did find that the default backingstore is rejected due to `verify auth failed` issues while it calls ping() and heartbeat() to noobaa core which can have an effect but I can't say for sure that it was the root cause.
